Add AWS Exporter to the docs (#211)
port-labs/port-docs: docs/build-your-software-catalog/sync-data-to-catalog/aws/run-on-events.md

              {
                "resource_type": "AWS::Lambda::Function",
                "requestFunctionName": "<requestFunctionName>",
                "responseFunctionName": "<responseFunctionName>",
                "eventName": "<eventName>",
                "region": "\"<awsRegion>\"",
                "identifier": "if .responseFunctionName != \"\" then .responseFunctionName else .requestFunctionName end",
                "action": "if .eventName | startswith(\"Delete\") then \"delete\" else \"upsert\" end"
              }
```

:::info
This example of CloudFormation stack can be used to act on changes of Lambda functions.
Moreover, it includes a parameter called `PortAWSExporterStackName`, that refers to the main exporter's stack name. That way, we can send the events to the existing exporter's events queue, for syncing relevant AWS resources.
:::

#### Properties

Let's review the properties of the event rule resource from above.

- The `EventBusName` property value is `default`, as this is the bus that gets events from various AWS services automatically:

```yaml showLineNumbers
Properties:
  # highlight-next-line
  EventBusName: default
  ...
```

> - The `EventPattern` property defines which events to consume, including which event sources, and other filters based on the event structure.

Reviewer: ```suggestion
- The `EventPattern` property defines which events to consume, including which event sources, which filters to apply based on the event structure.
```

  Here, we define the event pattern to consume Cloudtrail API calls, sourced from the Lambda service; On top of that, we filter only the `UpdateFunctionConfiguration`, `CreateFunction` and `DeleteFunction` events:

```yaml showLineNumbers
Properties:
  EventBusName: default
  # highlight-start
  EventPattern:
    source:
      - aws.lambda
    detail-type:
      - AWS API Call via CloudTrail
    detail:
      eventSource:
        - lambda.amazonaws.com
      eventName:
        - prefix: UpdateFunctionConfiguration
        - prefix: CreateFunction
        - prefix: DeleteFunction
  # highlight-end
```

:::tip
For composing an event pattern, learn more about events from AWS services, including different sources and event names [here](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-service-event.html).

For supported services in CloudTrail, click [here](https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-aws-service-specific-topics.html).
:::

- The `Targets` property defines the targets of the produced events.

>   In that case, we want to configure the AWS Exporter events queue as the target:

Reviewer: ```suggestion
 For the AWS exporter, we want to configure it's own event queue as the target:
```

>   In that case, we want to configure the AWS Exporter events queue as the target:

Reviewer: ```suggestion
 For the AWS exporter, we want to configure it's own event queue as the target:
```

```yaml showLineNumbers
Properties:
  ...
  # highlight-start
  Targets:
    - Id: "PortAWSExporterEventsQueue"
      Arn: { "Fn::ImportValue" : {"Fn::Sub": "${PortAWSExporterStackName}-EventsQueueARN" } }
      ...
  # highlight-end
```

> - The `InputTransformer` key in `Targets` property, defines the transformation applied to the event, before sending it to the target queue;

Reviewer: ```suggestion
- The `InputTransformer` key in the `Targets` property defines the transformation that will be applied to the event, before sending it to the target queue;
```

```yaml showLineNumbers
Properties:
  ...
  Targets:
    - Id: "PortAWSExporterEventsQueue"
      Arn: { "Fn::ImportValue" : {"Fn::Sub": "${PortAWSExporterStackName}-EventsQueueARN" } }
      # highlight-next-line
      InputTransformer:
        ...
```

> - `InputPathsMap` key in `InputTransformer`, defines which fields to extract from the event. It makes use of [`JSONPath`](https://goessner.net/articles/JsonPath/index.html) syntax to configure the extraction:

Reviewer: ```suggestion
- The `InputPathsMap` key in `InputTransformer` defines which fields to extract from the event. It makes use of [`JSONPath`](https://goessner.net/articles/JsonPath/index.html) syntax to configure the extraction:
```

```yaml showLineNumbers
Properties:
  ...
  Targets:
    - Id: "PortAWSExporterEventsQueue"
      Arn: { "Fn::ImportValue" : {"Fn::Sub": "${PortAWSExporterStackName}-EventsQueueARN" } }
      InputTransformer:
        # highlight-start
        InputPathsMap:
          awsRegion: $.detail.awsRegion
          eventName: $.detail.eventName
          requestFunctionName: $.detail.requestParameters.functionName
          responseFunctionName: $.detail.responseElements.functionName
        # highlight-end
        ...
```

For instance, the event received in EventBridge bus for the `CreateFunction` Lambda API call is something like:

```json showLineNumbers
{
  ...
  "detail": {
    ...
    "eventSource": "lambda.amazonaws.com",
    "eventName": "CreateFunction20150331",
    "awsRegion": "eu-west-1",
    ...
    "requestParameters": {
        "functionName": "test-function",
        ...
    },
    "responseElements": {
        "functionName": "test-function",
        ...
    },
    ...
  }
}
```

The `InputPathsMap` in the example, will extract only the `awsRegion`, `eventName`, and `functionName` from the API request and response.

> - `InputTemplate` key in `InputTransformer`, defines a template of the final input to the target queue, as the exporter's Lambda [expects](#input-structure-for-a-single-event) it. The template is rendered using the values from the `InputPathsMap`:

Reviewer: ```suggestion
- The `InputTemplate` key in `InputTransformer` defines a template of the final input that will be sent to the target queue, in the format the exporter's Lambda [expects](#input-structure-for-a-single-event). The template is rendered using the values from the `InputPathsMap`:
```

```yaml showLineNumbers
Properties:
  ...
  Targets:
    - Id: "PortAWSExporterEventsQueue"
      Arn: { "Fn::ImportValue" : {"Fn::Sub": "${PortAWSExporterStackName}-EventsQueueARN" } }
      InputTransformer:
        InputPathsMap:
          awsRegion: $.detail.awsRegion
          eventName: $.detail.eventName
          requestFunctionName: $.detail.requestParameters.functionName
          responseFunctionName: $.detail.responseElements.functionName
        # highlight-start
        InputTemplate: |-
          {
            "resource_type": "AWS::Lambda::Function",
            "requestFunctionName": "<requestFunctionName>",
            "responseFunctionName": "<responseFunctionName>",
            "eventName": "<eventName>",
            "region": "\"<awsRegion>\"",
            "identifier": "if .responseFunctionName != \"\" then .responseFunctionName else .requestFunctionName end",
            "action": "if .eventName | startswith(\"Delete\") then \"delete\" else \"upsert\" end"
          }
        # highlight-end
```

> Here, the `InputTemplate` takes the information extracted in the `InputPathsMap`, and constructs a message that will be sent to the SQS queue, and will be consumed by the Exporter’s Lambda.

Reviewer: ```suggestion
Here, the `InputTemplate` takes the information extracted in the `InputPathsMap`, and constructs a message that will be sent to the SQS queue and consumed by the Exporter’s Lambda.
```

> The message includes the hardcoded resource type (`AWS::Lambda::Function`); the `awsRegion` from the event;

Reviewer: Update this section so that it is a proper list of resources sent, and not a very long sentence with semi-colons in between

a JQ query to figure out the action (`upsert` or `delete` the Port entity), based on the `eventName`; and a JQ query to calculate the identifier, based on the function name from the API response or request.

:::info
To read more about input transformer in EventBridge, click [here](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-transform-target-input.html).
:::

## Setup summary

> 1. You should prepare an [event rule](#event-rule-definition), based on any specific events;

Reviewer: ```suggestion
1. You should prepare an [event rule](#event-rule-definition), based on specific events matching resources you want the AWS exporter to update in real time;
```

[The AWS exporter application](./aws.md#exporter-aws-serverless-application) creates an SQS Queue as part of its installation process, and configures it as an event source for the exporter's lambda.

This allows you to set up an event rule, that consumes any events from the bus and sends them to the AWS exporter's queue.

Each event will have to be transformed as part of the event rule, before reaching the queue, to suit the exporter's lambda expected event structure.
The transformation goal is to map an event to a single AWS resource, that the exporter can handle.

The events in the queue are consumed automatically by the exporter's lambda, that will sync the updated state of the relevant AWS resources with Port.

### Input structure for a single event

The AWS exporter's lambda accepts JSON events with the following structure:

- `resource_type` - a hardcoded string representing the AWS resource type that is configured in the [`config.json`](./aws.md#exporter-configjson-file) of the AWS Exporter;

> - `region` - A JQ query to get the region from the event, usually the value will be `"\"<awsRegion>\""`

Reviewer 2: ```suggestion
- `region` - A JQ query to get the region from the event, usually the value will be `"\"<awsRegion>\""`;
```
